### PR TITLE
Introduce CanvasSectionContainer class.

### PR DESCRIPTION
**Pull request by gokaysatir**
Signed-off-by: Gökay Şatır <gokaysatir@collabora.com>
Change-Id: I69929595450c48ed45326c21743fbde3babc9803

* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: loleaflet/Makefile.am [hunk 1]
```diff
@@ -177,6 +177,11 @@ NODE_MODULES_JS_DST = $(patsubst %.js,$(DIST_FOLDER)/%.js,$(NODE_MODULES_JS))
 LOLEAFLET_LIBS_JS_SRC = $(patsubst %.js,$(srcdir)/js/%.js,$(LOLEAFLET_LIBS_JS))
 LOLEAFLET_LIBS_JS_DST = $(patsubst %.js,$(DIST_FOLDER)/%.js,$(LOLEAFLET_LIBS_JS))
 
+CANVASSECTION:
+	$(srcdir)/node_modules/typescript/bin/tsc --project $(srcdir)/tsconfig.json
```
Comment (dennisfrancis) [inline loleaflet/Makefile.am:181]: One suggestion : Is it possible to write the js file directly to the DIST_FOLDER/<...>/ here via tsconfig.json ? so that we can avoid the 'copy' steps (assuming the js linter will not complain because of this).

**Review by dennisfrancis — COMMENTED** [review]
I'm not sure why this does not work with one make call. It looks very much like the tsc call happens after the copy is attempted. I've made one suggestion inline.

File: loleaflet/Makefile.am (hunk 1 above)
Comment (gokaysatir) [reply loleaflet/Makefile.am:181]: Thank you for the suggestion, i will try it asap :)

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 2]
```diff
@@ -349,10 +351,6 @@ L.CanvasTilePainter = L.Class.extend({
 		var partChanged = (part !== this._lastPart);
 
 		var mapSizeChanged = !newMapSize.equals(this._lastMapSize);
-		// To avoid flicker, only resize the canvas element if width or height of the map increases.
-		var newSize = new L.Point(Math.max(newMapSize.x, this._lastSize ? this._lastSize.x : 0),
-					  Math.max(newMapSize.y, this._lastSize ? this._lastSize.y : 0));
-		var resizeCanvas = !this._lastSize || !newSize.equals(this._lastSize);
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:355]: If I remember correctly this was used to avoid flicker when changing to/from cell edit-mode in a device with on-screen keyboard and also during sidebar switch in desktop browser. It would be good to check for this behavior.

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 3]
```diff
@@ -352,181 +399,29 @@ L.TileSectionManager = L.Class.extend({
 	_updateWithRAF: function () {
 		// update-loop with requestAnimationFrame
 		this._canvasRAF = L.Util.requestAnimFrame(this._updateWithRAF, this, false /* immediate */);
-		this.update();
+		this._sectionContainer.requestReDraw();
 	},
 
 	update: function () {
-
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:330]: Are the below checks done somewhere before attempting a full redraw ? These checks help us to avoid unnecessary redraws etc when these are done via RAF loop which is good w.r.t. cpu/battery.

File: loleaflet/src/layer/tile/CanvasTileLayer.js (hunk 2 above)
Comment (gokaysatir) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:355]: I checked them now, they seem good.

File: loleaflet/src/layer/tile/CanvasTileLayer.js (hunk 3 above)
Comment (gokaysatir) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:330]: All things were checked inside update function. Now appropriate callbacks are called when an event occurs. So, we don't need them now.

I also saw that loop may need some checks but while moving, there will be changes most probably.

Thanks for the comments:)

**Comment by gokaysatir** [comment]
@dennisfrancis, i rebased your commits with the ones using canvas sections. Can i have your opinion? Thank you :)

File: loleaflet/Makefile.am [hunk 4]
```diff
@@ -28,9 +28,9 @@ $(L10N_IOS_ALL_JS) : $(wildcard $(srcdir)/po/ui-*.po) $(shell find $(srcdir)/l10
 endif
 
 SRC_TS = $(wildcard src/layer/tile/*.ts)
-SRC_TS_JS = $(patsubst src/layer/tile/%.ts,$(DIST_FOLDER)/src/layer/tile/%.js,$(SRC_TS))
```
Comment (dennisfrancis) [inline loleaflet/Makefile.am]: Shouldn't the changes to Makefile.am be part of the first commit where the .ts file was added ? ("git rebase -i" can fix this)

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 5]
```diff
@@ -231,12 +232,12 @@ L.TileSectionManager = L.Class.extend({
 						this._zoomAnimation();
 					}
 				},
-				zoomStepEnd: function () {
+				zoomStepEnd: function (zoom, newCenter) {
```
Comment (dennisfrancis) [inline loleaflet/src/layer/tile/CanvasTileLayer.js:235]: The changes in this file also seems unrelated to "Control.Header: Bring non-shared functions from rows and columns". Either interactive rebase can be used to move this to relevant previous commit or this file's change can be made into a separate commit.

**Review by dennisfrancis — COMMENTED** [review]
Looks good apart from some rebase related issues on the last commit (commented on that).

File: loleaflet/src/layer/tile/CanvasTileLayer.js (hunk 5 above)
Comment (dennisfrancis) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:235]: My comments on rebasing are relevant if the commit history in this PR is to be maintained when merging to master.

**Review by dennisfrancis — COMMENTED** [review]
Looks great, I have a couple of small comments which unfortunately I did not see while reviewing earlier. Thanks!

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 6]
```diff
@@ -264,13 +175,13 @@ L.TileSectionManager = L.Class.extend({
...
 
-				this._drawTileInPane(tile, tileBounds, paneBounds, paneOffset, this._canvasCtx);
+				this._drawTileInPane(tile, tileBounds, paneBounds, paneOffset, this._tilesSection.context);
 			}
 
 			if (extendedBounds.intersects(tileBounds)) {
 				var offset = extendedBounds.getTopLeft();
-				offset.x = Math.min(offset.x, viewBounds.min.x);
-				offset.y = Math.min(offset.y, viewBounds.min.y);
+				viewBounds.x = Math.min(offset.x, viewBounds.min.x);
```
[the remark this answers is not in the document]
Comment (gokaysatir) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:157]: You are right, i think i accidentally renamed those lines while i rebase commits.

File: loleaflet/src/layer/tile/CanvasTileLayer.js [hunk 7]
```diff
@@ -320,9 +231,39 @@ L.TileSectionManager = L.Class.extend({
...
+			size: [0, 0], // Going to be expanded, no initial width or height is necessary.
+			expand: 'top left bottom right', // Expand to all directions.
+			drawingOrder: 1,
+			zIndex: 5,
+			interactable: true,
+			myProperties: {
+				docLayer: that._layer,
+				tsManager: that
+			},
+			onInitialize: that._onTilesSectionInitialize,
```
[the remark this answers is not in the document]
Comment (gokaysatir) [reply loleaflet/src/layer/tile/CanvasTileLayer.js:225]: In fact i tried it, but when a section's function needs to call another function inside the same section, things get messy. I hope that problem can easily be seen by checking the contents of "this" variable while debugging. If we can get used to this usage, we can think functions as if they were defined inside section definition. Perhaps that maybe an advantage while we think on the code, in terms of understanding & designing.
